hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java

```diff
@@ -172,13 +172,55 @@ public LoaderContext getLoaderContext() {
       return ctx;
     }
 
+    /***
+     * a thread for parallel MD5 computing to increase performance when loading
+     */
+    private static class DigestThread extends Thread {
+      volatile private IOException ioe = null;
+      volatile private MD5Hash digest = null;
+      private File file;
+
+      public DigestThread(File inFile) {
+        file = inFile;
+      }
+
+      public MD5Hash getDigest() {
+        return digest;
+      }
+
+      public IOException getException() {
+        return ioe;
+      }
+
+      @Override
+      public void run() {
+        try {
+          digest = MD5FileUtils.computeMd5ForFile(file);
+        } catch (IOException e) {
+          ioe = e;
+        } catch (Throwable t) {
+          ioe = new IOException(t);
+        }
+      }
+    }
+
     void load(File file) throws IOException {
       long start = Time.monotonicNow();
-      imgDigest = MD5FileUtils.computeMd5ForFile(file);
+      DigestThread dt = new DigestThread(file);
+      dt.start();
       RandomAccessFile raFile = new RandomAccessFile(file, "r");
       FileInputStream fin = new FileInputStream(file);
       try {
         loadInternal(raFile, fin);
+        try {
+          dt.join();
+          if (dt.getException() != null) {
+            throw dt.getException();
+          }
+          imgDigest = dt.getDigest();
```

Comment:
If you do the approach of getDigest to throw the exception, you cover this.

Reply:
this exception is throwed when call imgDigest = MD5FileUtils.computeMd5ForFile(file);
according to digest calculation and load code，this exception is also directly throwed. So I do it.

Reply:
My point is following the other comment.
If getDigest checks if there is an exception and throws it here, you don't need to check for the exception here.
You can also remove dt.getException().